pip install pandas

!pip install pandas


In [1]:
import pandas as pd

* Wczytać `AiSD2.csv` przez `pd.read_csv()`
* Odwołać się do kolumny `Nazwisko` (wyświetlić ją). Czy się udało?

In [2]:
oceny = pd.read_csv("AiSD2.csv")
oceny.head()

Imię         Nazwisko                Z1(4+1)    Z2(4+1)      Z3(4+1)  \
0      Anna            Nowak                2.5+1.0    4.0+1.0      2.0+0.0   
1     Rafał         Kowalski                3.0+0.0    2.0+0.5      2.0+1.0   
2     Pavlo       Wiśniewski                2.5+0.5     US+1.0      2.5+1.0   
3    Aleksy        Dąbrowski                0.5+1.0    1.0+1.0        NA+NA   
4  Jarosław      Lewandowski                2.0+0.0    3.0+1.0      1.0+0.0   

      Z4(4+1)     Z5(2+2)      Z6(4+1)     Z7(4+1)     Z8(4+1)     Z9(4+1)  \
0     1.0+2.0     1.5+2.0      2.0+0.0      0.0+NB     1.5+1.0     3.5+0.5   
1     1.5+1.5     2.0+1.0      1.0+0.5      0.0+NB     3.0+1.0     3.5+1.0   
2     1.5+1.5     2.0+1.5      3.5+0.0     1.0+1.0     3.0+1.0     3.5+1.0   
3       NA+NA       NA+NA        NA+NA       NA+NA       NA+NA       NA+NA   
4     1.5+1.5     0.5+0.0     1.25+0.0      0.0+NB     2.5+0.0       NB+NB   

     Z10(4+1)      Z11(4+1)              Z12(4+1)                  Z13(5)  \
0     2.0+0.0       0.5+1.0                 ZW+ZW                 4.0+0.0   
1      0.0+NB       1.0+1.0     NS.(3.0)+NS.(1.0)                 4.0+0.0   
2       ZW+ZW       1.0+0.5               3.5+1.0                 4.0+1.0   
3       NA+NA         NA+NA                 NA+NA                   NA+NA   
4       NB+NB         NB+NB                 NB+NB                   NB+NB   

    Z14(4+1) Z15(5)  
0    4.0+1.0     NA  
1    4.0+1.0     NA  
2    4.0+1.0     NA  
3      NA+NA     NA  
4      NB+NB     NA

In [3]:
oceny.Nazwisko

AttributeError: 'DataFrame' object has no attribute 'Nazwisko'

In [4]:
oceny["Nazwisko"]

KeyError: 'Nazwisko'

In [5]:
oceny.columns

Index(['Imię', '      Nazwisko', '           Z1(4+1)', '  Z2(4+1)',
       '    Z3(4+1)', '   Z4(4+1)', '   Z5(2+2)', '   Z6(4+1)', '  Z7(4+1)',
       '   Z8(4+1)', '   Z9(4+1)', ' Z10(4+1)', '    Z11(4+1)', '  Z12(4+1)',
       '            Z13(5)', '  Z14(4+1)', 'Z15(5)'],
      dtype='object')

In [6]:
oceny["      Nazwisko"]

0               Nowak
1            Kowalski
2          Wiśniewski
3           Dąbrowski
4         Lewandowski
5              Wójcik
6            Kamiński
7           Kowalczyk
8           Zielińska
9           Szymański
10            Woźniak
Name:       Nazwisko, dtype: object

In [7]:
oceny[["      Nazwisko", "   Z4(4+1)"]]

Nazwisko     Z4(4+1)
0             Nowak     1.0+2.0
1          Kowalski     1.5+1.5
2        Wiśniewski     1.5+1.5
3         Dąbrowski       NA+NA
4       Lewandowski     1.5+1.5
5            Wójcik     1.5+1.5
6          Kamiński     1.5+0.5
7         Kowalczyk     4.0+2.0
8         Zielińska     1.0+0.0
9         Szymański     0.0+1.5
10          Woźniak     3.0+2.0

In [9]:
oceny["      Nazwisko"].values

array(['      Nowak', '     Kowalski', '     Wiśniewski', '    Dąbrowski',
       '  Lewandowski', 'Wójcik', '    Kamiński', '     Kowalczyk',
       '    Zielińska', '   Szymański', '    Woźniak'], dtype=object)

Jak to rozwiązać?

* na piechotę (stip(), apply())
* elegancka metoda, na etapie wczytywania

## Na piechotę

In [13]:
oceny.columns.str.strip()

Index(['Imię', 'Nazwisko', 'Z1(4+1)', 'Z2(4+1)', 'Z3(4+1)', 'Z4(4+1)',
       'Z5(2+2)', 'Z6(4+1)', 'Z7(4+1)', 'Z8(4+1)', 'Z9(4+1)', 'Z10(4+1)',
       'Z11(4+1)', 'Z12(4+1)', 'Z13(5)', 'Z14(4+1)', 'Z15(5)'],
      dtype='object')

In [14]:
oceny.columns

Index(['Imię', '      Nazwisko', '           Z1(4+1)', '  Z2(4+1)',
       '    Z3(4+1)', '   Z4(4+1)', '   Z5(2+2)', '   Z6(4+1)', '  Z7(4+1)',
       '   Z8(4+1)', '   Z9(4+1)', ' Z10(4+1)', '    Z11(4+1)', '  Z12(4+1)',
       '            Z13(5)', '  Z14(4+1)', 'Z15(5)'],
      dtype='object')

In [18]:
oceny.columns = oceny.columns.str.strip() # to nie naprawia wartości w wierszach

In [16]:
oceny.columns

Index(['Imię', 'Nazwisko', 'Z1(4+1)', 'Z2(4+1)', 'Z3(4+1)', 'Z4(4+1)',
       'Z5(2+2)', 'Z6(4+1)', 'Z7(4+1)', 'Z8(4+1)', 'Z9(4+1)', 'Z10(4+1)',
       'Z11(4+1)', 'Z12(4+1)', 'Z13(5)', 'Z14(4+1)', 'Z15(5)'],
      dtype='object')

In [17]:
oceny.Nazwisko

0               Nowak
1            Kowalski
2          Wiśniewski
3           Dąbrowski
4         Lewandowski
5              Wójcik
6            Kamiński
7           Kowalczyk
8           Zielińska
9           Szymański
10            Woźniak
Name: Nazwisko, dtype: object

In [19]:
oceny.Nazwisko.values

array(['      Nowak', '     Kowalski', '     Wiśniewski', '    Dąbrowski',
       '  Lewandowski', 'Wójcik', '    Kamiński', '     Kowalczyk',
       '    Zielińska', '   Szymański', '    Woźniak'], dtype=object)

In [21]:
oceny.Nazwisko.str.strip().values

array(['Nowak', 'Kowalski', 'Wiśniewski', 'Dąbrowski', 'Lewandowski',
       'Wójcik', 'Kamiński', 'Kowalczyk', 'Zielińska', 'Szymański',
       'Woźniak'], dtype=object)

In [22]:
oceny.Nazwisko.values

array(['      Nowak', '     Kowalski', '     Wiśniewski', '    Dąbrowski',
       '  Lewandowski', 'Wójcik', '    Kamiński', '     Kowalczyk',
       '    Zielińska', '   Szymański', '    Woźniak'], dtype=object)

In [24]:
oceny.str.strip() # niestety nie ma

AttributeError: 'DataFrame' object has no attribute 'str'

In [25]:
oceny.strip() # niestety nie ma

AttributeError: 'DataFrame' object has no attribute 'strip'

In [ ]:
oceny.Nazwisko.str.strip()

In [31]:
oceny2 = oceny.apply(lambda x: x.str.strip()) # tu mamy trochę szczęścia że zadziałało

In [32]:
oceny2.head()

Imię     Nazwisko  Z1(4+1)  Z2(4+1)  Z3(4+1)  Z4(4+1)  Z5(2+2)  \
0      Anna        Nowak  2.5+1.0  4.0+1.0  2.0+0.0  1.0+2.0  1.5+2.0   
1     Rafał     Kowalski  3.0+0.0  2.0+0.5  2.0+1.0  1.5+1.5  2.0+1.0   
2     Pavlo   Wiśniewski  2.5+0.5   US+1.0  2.5+1.0  1.5+1.5  2.0+1.5   
3    Aleksy    Dąbrowski  0.5+1.0  1.0+1.0    NA+NA    NA+NA    NA+NA   
4  Jarosław  Lewandowski  2.0+0.0  3.0+1.0  1.0+0.0  1.5+1.5  0.5+0.0   

    Z6(4+1)  Z7(4+1)  Z8(4+1)  Z9(4+1) Z10(4+1) Z11(4+1)           Z12(4+1)  \
0   2.0+0.0   0.0+NB  1.5+1.0  3.5+0.5  2.0+0.0  0.5+1.0              ZW+ZW   
1   1.0+0.5   0.0+NB  3.0+1.0  3.5+1.0   0.0+NB  1.0+1.0  NS.(3.0)+NS.(1.0)   
2   3.5+0.0  1.0+1.0  3.0+1.0  3.5+1.0    ZW+ZW  1.0+0.5            3.5+1.0   
3     NA+NA    NA+NA    NA+NA    NA+NA    NA+NA    NA+NA              NA+NA   
4  1.25+0.0   0.0+NB  2.5+0.0    NB+NB    NB+NB    NB+NB              NB+NB   

    Z13(5) Z14(4+1) Z15(5)  
0  4.0+0.0  4.0+1.0     NA  
1  4.0+0.0  4.0+1.0     NA  
2  4.0+1.0  4.0+1.0     NA  
3    NA+NA    NA+NA     NA  
4    NB+NB    NB+NB     NA

In [33]:
oceny2["Z4(4+1)"].values

array(['1.0+2.0', '1.5+1.5', '1.5+1.5', 'NA+NA', '1.5+1.5', '1.5+1.5',
       '1.5+0.5', '4.0+2.0', '1.0+0.0', '0.0+1.5', '3.0+2.0'],
      dtype=object)

In [34]:
oceny.dtypes

Imię        object
Nazwisko    object
Z1(4+1)     object
Z2(4+1)     object
Z3(4+1)     object
Z4(4+1)     object
Z5(2+2)     object
Z6(4+1)     object
Z7(4+1)     object
Z8(4+1)     object
Z9(4+1)     object
Z10(4+1)    object
Z11(4+1)    object
Z12(4+1)    object
Z13(5)      object
Z14(4+1)    object
Z15(5)      object
dtype: object

Dygresja: czym się różnią metody `apply()` oraz `applymap()` i na czym operują?

In [35]:
oceny.apply(lambda x: print(x))

0           Anna
1          Rafał
2          Pavlo
3         Aleksy
4       Jarosław
5     Bartłomiej
6         Łukasz
7          Oskar
8         Joanna
9        Oktawia
10        Patryk
Name: Imię, dtype: object
0               Nowak
1            Kowalski
2          Wiśniewski
3           Dąbrowski
4         Lewandowski
5              Wójcik
6            Kamiński
7           Kowalczyk
8           Zielińska
9           Szymański
10            Woźniak
Name: Nazwisko, dtype: object
0                   2.5+1.0
1                   3.0+0.0
2                   2.5+0.5
3                   0.5+1.0
4                   2.0+0.0
5                   2.0+1.0
6                   4.0+1.0
7                   4.0+0.5
8                   2.5+0.0
9                   2.5+1.0
10                  2.0+1.0
Name: Z1(4+1), dtype: object
0       4.0+1.0
1       2.0+0.5
2        US+1.0
3       1.0+1.0
4       3.0+1.0
5       2.0+1.0
6       2.0+1.0
7       4.0+1.0
8         US+US
9       0.0+0.0
10      4.0+1.0
Na

Imię        None
Nazwisko    None
Z1(4+1)     None
Z2(4+1)     None
Z3(4+1)     None
Z4(4+1)     None
Z5(2+2)     None
Z6(4+1)     None
Z7(4+1)     None
Z8(4+1)     None
Z9(4+1)     None
Z10(4+1)    None
Z11(4+1)    None
Z12(4+1)    None
Z13(5)      None
Z14(4+1)    None
Z15(5)      None
dtype: object

In [36]:
oceny.apply(lambda x: print(type(x)))

<class 'pandas.core.series.Series'>
<class 'pandas.core.series.Series'>
<class 'pandas.core.series.Series'>
<class 'pandas.core.series.Series'>
<class 'pandas.core.series.Series'>
<class 'pandas.core.series.Series'>
<class 'pandas.core.series.Series'>
<class 'pandas.core.series.Series'>
<class 'pandas.core.series.Series'>
<class 'pandas.core.series.Series'>
<class 'pandas.core.series.Series'>
<class 'pandas.core.series.Series'>
<class 'pandas.core.series.Series'>
<class 'pandas.core.series.Series'>
<class 'pandas.core.series.Series'>
<class 'pandas.core.series.Series'>
<class 'pandas.core.series.Series'>


Imię        None
Nazwisko    None
Z1(4+1)     None
Z2(4+1)     None
Z3(4+1)     None
Z4(4+1)     None
Z5(2+2)     None
Z6(4+1)     None
Z7(4+1)     None
Z8(4+1)     None
Z9(4+1)     None
Z10(4+1)    None
Z11(4+1)    None
Z12(4+1)    None
Z13(5)      None
Z14(4+1)    None
Z15(5)      None
dtype: object

In [38]:
oceny.apply(lambda x: print(x), axis=1) # axis kontroluje czy to wiersze czy kolumny, tutaj to są wiersze

Imię                          Anna
Nazwisko                     Nowak
Z1(4+1)                    2.5+1.0
Z2(4+1)                    4.0+1.0
Z3(4+1)                    2.0+0.0
Z4(4+1)                    1.0+2.0
Z5(2+2)                    1.5+2.0
Z6(4+1)                    2.0+0.0
Z7(4+1)                     0.0+NB
Z8(4+1)                    1.5+1.0
Z9(4+1)                    3.5+0.5
Z10(4+1)                   2.0+0.0
Z11(4+1)                   0.5+1.0
Z12(4+1)                     ZW+ZW
Z13(5)                     4.0+0.0
Z14(4+1)                   4.0+1.0
Z15(5)                          NA
Name: 0, dtype: object
Imię                        Rafał
Nazwisko                 Kowalski
Z1(4+1)                   3.0+0.0
Z2(4+1)                   2.0+0.5
Z3(4+1)                   2.0+1.0
Z4(4+1)                   1.5+1.5
Z5(2+2)                   2.0+1.0
Z6(4+1)                   1.0+0.5
Z7(4+1)                    0.0+NB
Z8(4+1)                   3.0+1.0
Z9(4+1)                   3.5+1.0
Z10(4+1)

0     None
1     None
2     None
3     None
4     None
5     None
6     None
7     None
8     None
9     None
10    None
dtype: object

In [42]:
oceny.applymap(lambda x: print(x)) # działa na pojedyńczych komórkach

Anna
Rafał
Pavlo
Aleksy
Jarosław
Bartłomiej
Łukasz
Oskar
Joanna
Oktawia
Patryk
      Nowak
     Kowalski
     Wiśniewski
    Dąbrowski
  Lewandowski
Wójcik
    Kamiński
     Kowalczyk
    Zielińska
   Szymański
    Woźniak
              2.5+1.0
              3.0+0.0
      2.5+0.5
            0.5+1.0
    2.0+0.0
    2.0+1.0
    4.0+1.0
              4.0+0.5
      2.5+0.0
           2.5+1.0
            2.0+1.0
  4.0+1.0
  2.0+0.5
  US+1.0
  1.0+1.0
  3.0+1.0
  2.0+1.0
  2.0+1.0
  4.0+1.0
  US+US
  0.0+0.0
  4.0+1.0
    2.0+0.0
    2.0+1.0
    2.5+1.0
    NA+NA
    1.0+0.0
    2.5+0.5
    3.5+1.0
    2.0+1.0
    0.5+0.0
    1.0+0.5
    2.5+1.0
   1.0+2.0
   1.5+1.5
   1.5+1.5
   NA+NA
   1.5+1.5
   1.5+1.5
   1.5+0.5
   4.0+2.0
   1.0+0.0
   0.0+1.5
   3.0+2.0
   1.5+2.0
   2.0+1.0
   2.0+1.5
   NA+NA
   0.5+0.0
   0.5+1.0
   1.5+0.5
   2.0+1.0
   0.5+1.0
   0.5+0.5
   1.0+0.5
   2.0+0.0
   1.0+0.5
   3.5+0.0
   NA+NA
   1.25+0.0
   1.75+0.5
   US+1.0
   1.5+0.0
   1.0+0.0
   NB+NB
   1.0

Imię Nazwisko Z1(4+1) Z2(4+1) Z3(4+1) Z4(4+1) Z5(2+2) Z6(4+1) Z7(4+1)  \
0   None     None    None    None    None    None    None    None    None   
1   None     None    None    None    None    None    None    None    None   
2   None     None    None    None    None    None    None    None    None   
3   None     None    None    None    None    None    None    None    None   
4   None     None    None    None    None    None    None    None    None   
5   None     None    None    None    None    None    None    None    None   
6   None     None    None    None    None    None    None    None    None   
7   None     None    None    None    None    None    None    None    None   
8   None     None    None    None    None    None    None    None    None   
9   None     None    None    None    None    None    None    None    None   
10  None     None    None    None    None    None    None    None    None   

   Z8(4+1) Z9(4+1) Z10(4+1) Z11(4+1) Z12(4+1) Z13(5) Z14(4+1) Z15(5)  
0     None    None     None     None     None   None     None   None  
1     None    None     None     None     None   None     None   None  
2     None    None     None     None     None   None     None   None  
3     None    None     None     None     None   None     None   None  
4     None    None     None     None     None   None     None   None  
5     None    None     None     None     None   None     None   None  
6     None    None     None     None     None   None     None   None  
7     None    None     None     None     None   None     None   None  
8     None    None     None     None     None   None     None   None  
9     None    None     None     None     None   None     None   None  
10    None    None     None     None     None   None     None   None

Jak zastosować stripa typlko na kolumnach pewnego typu?

In [45]:
x = 11
zmienna = 22 if x > 10 else 33

In [46]:
zmienna

22

In [47]:
if x > 10:
    zmienna = 22
else:
    zmienna = 33
zmienna

22

In [48]:
x = "     Janek   "

In [49]:
"     Janek   ".strip()

'Janek'

In [50]:
zmienna = x.strip() if isinstance(x, str) else x

In [51]:
zmienna

'Janek'

In [52]:
x = 6

In [53]:
zmienna = x.strip() if isinstance(x, str) else x

In [54]:
zmienna

6

In [55]:
x.strip()

AttributeError: 'int' object has no attribute 'strip'

In [60]:
df = oceny.applymap(lambda pojedynczy_napis: pojedynczy_napis.strip() if isinstance(pojedynczy_napis, str) else pojedynczy_napis)

In [61]:
df.Nazwisko.values

array(['Nowak', 'Kowalski', 'Wiśniewski', 'Dąbrowski', 'Lewandowski',
       'Wójcik', 'Kamiński', 'Kowalczyk', 'Zielińska', 'Szymański',
       'Woźniak'], dtype=object)

In [62]:
oceny.applymap(lambda pojedynczy_napis: pojedynczy_napis.strip() if isinstance(pojedynczy_napis, str) else pojedynczy_napis)

Imię     Nazwisko  Z1(4+1)  Z2(4+1)  Z3(4+1)  Z4(4+1)  Z5(2+2)  \
0         Anna        Nowak  2.5+1.0  4.0+1.0  2.0+0.0  1.0+2.0  1.5+2.0   
1        Rafał     Kowalski  3.0+0.0  2.0+0.5  2.0+1.0  1.5+1.5  2.0+1.0   
2        Pavlo   Wiśniewski  2.5+0.5   US+1.0  2.5+1.0  1.5+1.5  2.0+1.5   
3       Aleksy    Dąbrowski  0.5+1.0  1.0+1.0    NA+NA    NA+NA    NA+NA   
4     Jarosław  Lewandowski  2.0+0.0  3.0+1.0  1.0+0.0  1.5+1.5  0.5+0.0   
5   Bartłomiej       Wójcik  2.0+1.0  2.0+1.0  2.5+0.5  1.5+1.5  0.5+1.0   
6       Łukasz     Kamiński  4.0+1.0  2.0+1.0  3.5+1.0  1.5+0.5  1.5+0.5   
7        Oskar    Kowalczyk  4.0+0.5  4.0+1.0  2.0+1.0  4.0+2.0  2.0+1.0   
8       Joanna    Zielińska  2.5+0.0    US+US  0.5+0.0  1.0+0.0  0.5+1.0   
9      Oktawia    Szymański  2.5+1.0  0.0+0.0  1.0+0.5  0.0+1.5  0.5+0.5   
10      Patryk      Woźniak  2.0+1.0  4.0+1.0  2.5+1.0  3.0+2.0  1.0+0.5   

     Z6(4+1)  Z7(4+1)  Z8(4+1)  Z9(4+1) Z10(4+1) Z11(4+1)           Z12(4+1)  \
0    2.0+0.0   0.0+NB  1.5+1.0  3.5+0.5  2.0+0.0  0.5+1.0              ZW+ZW   
1    1.0+0.5   0.0+NB  3.0+1.0  3.5+1.0   0.0+NB  1.0+1.0  NS.(3.0)+NS.(1.0)   
2    3.5+0.0  1.0+1.0  3.0+1.0  3.5+1.0    ZW+ZW  1.0+0.5            3.5+1.0   
3      NA+NA    NA+NA    NA+NA    NA+NA    NA+NA    NA+NA              NA+NA   
4   1.25+0.0   0.0+NB  2.5+0.0    NB+NB    NB+NB    NB+NB              NB+NB   
5   1.75+0.5  0.0+1.0  1.5+1.0  4.0+1.0    ZW+ZW  0.5+1.0            3.0+0.0   
6     US+1.0  0.5+1.0  2.0+1.0  3.5+1.0  2.0+0.0  0.5+0.5            3.0+1.0   
7    1.5+0.0  1.0+1.0  4.0+1.0  4.0+1.0  1.0+0.5   0.0+NB            3.5+1.0   
8    1.0+0.0   0.5+NB    NB+NB    NB+NB    NB+NB    NB+NB              NB+NB   
9      NB+NB   0.0+NB    NB+NB    NB+NB    NB+NB    NB+NB              NB+NB   
10   1.0+0.5  1.0+0.5  2.0+1.0  4.0+1.0  2.0+0.5  4.0+1.0            4.0+1.0   

     Z13(5) Z14(4+1) Z15(5)  
0   4.0+0.0  4.0+1.0     NA  
1   4.0+0.0  4.0+1.0     NA  
2   4.0+1.0  4.0+1.0     NA  
3     NA+NA    NA+NA     NA  
4     NB+NB    NB+NB     NA  
5   4.0+1.0  3.5+0.0     NA  
6   2.0+1.0  4.0+1.0     NA  
7   4.0+1.0  3.0+0.0     NA  
8     NB+NB    NB+NB     NA  
9     NB+NB    NB+NB     NA  
10  2.0+0.0  4.0+1.0     NA

In [70]:
oceny["float"] = 1.0
oceny["int"] = 1

In [71]:
oceny.head()

Imię         Nazwisko                Z1(4+1)    Z2(4+1)      Z3(4+1)  \
0      Anna            Nowak                2.5+1.0    4.0+1.0      2.0+0.0   
1     Rafał         Kowalski                3.0+0.0    2.0+0.5      2.0+1.0   
2     Pavlo       Wiśniewski                2.5+0.5     US+1.0      2.5+1.0   
3    Aleksy        Dąbrowski                0.5+1.0    1.0+1.0        NA+NA   
4  Jarosław      Lewandowski                2.0+0.0    3.0+1.0      1.0+0.0   

      Z4(4+1)     Z5(2+2)      Z6(4+1)     Z7(4+1)     Z8(4+1)     Z9(4+1)  \
0     1.0+2.0     1.5+2.0      2.0+0.0      0.0+NB     1.5+1.0     3.5+0.5   
1     1.5+1.5     2.0+1.0      1.0+0.5      0.0+NB     3.0+1.0     3.5+1.0   
2     1.5+1.5     2.0+1.5      3.5+0.0     1.0+1.0     3.0+1.0     3.5+1.0   
3       NA+NA       NA+NA        NA+NA       NA+NA       NA+NA       NA+NA   
4     1.5+1.5     0.5+0.0     1.25+0.0      0.0+NB     2.5+0.0       NB+NB   

     Z10(4+1)      Z11(4+1)              Z12(4+1)                  Z13(5)  \
0     2.0+0.0       0.5+1.0                 ZW+ZW                 4.0+0.0   
1      0.0+NB       1.0+1.0     NS.(3.0)+NS.(1.0)                 4.0+0.0   
2       ZW+ZW       1.0+0.5               3.5+1.0                 4.0+1.0   
3       NA+NA         NA+NA                 NA+NA                   NA+NA   
4       NB+NB         NB+NB                 NB+NB                   NB+NB   

   Z14(4+1) Z15(5)  float  int  
0   4.0+1.0     NA    1.0    1  
1   4.0+1.0     NA    1.0    1  
2   4.0+1.0     NA    1.0    1  
3     NA+NA     NA    1.0    1  
4     NB+NB     NA    1.0    1

In [72]:
oceny.dtypes

Imię         object
Nazwisko     object
Z1(4+1)      object
Z2(4+1)      object
Z3(4+1)      object
Z4(4+1)      object
Z5(2+2)      object
Z6(4+1)      object
Z7(4+1)      object
Z8(4+1)      object
Z9(4+1)      object
Z10(4+1)     object
Z11(4+1)     object
Z12(4+1)     object
Z13(5)       object
Z14(4+1)     object
Z15(5)       object
float       float64
int           int64
dtype: object

In [77]:
oceny_obj = oceny.select_dtypes(['object'])

In [78]:
oceny_obj

Imię         Nazwisko                Z1(4+1)    Z2(4+1)  \
0         Anna            Nowak                2.5+1.0    4.0+1.0   
1        Rafał         Kowalski                3.0+0.0    2.0+0.5   
2        Pavlo       Wiśniewski                2.5+0.5     US+1.0   
3       Aleksy        Dąbrowski                0.5+1.0    1.0+1.0   
4     Jarosław      Lewandowski                2.0+0.0    3.0+1.0   
5   Bartłomiej           Wójcik                2.0+1.0    2.0+1.0   
6       Łukasz         Kamiński                4.0+1.0    2.0+1.0   
7        Oskar        Kowalczyk                4.0+0.5    4.0+1.0   
8       Joanna        Zielińska                2.5+0.0      US+US   
9      Oktawia        Szymański                2.5+1.0    0.0+0.0   
10      Patryk          Woźniak                2.0+1.0    4.0+1.0   

        Z3(4+1)     Z4(4+1)     Z5(2+2)      Z6(4+1)      Z7(4+1)     Z8(4+1)  \
0       2.0+0.0     1.0+2.0     1.5+2.0      2.0+0.0       0.0+NB     1.5+1.0   
1       2.0+1.0     1.5+1.5     2.0+1.0      1.0+0.5       0.0+NB     3.0+1.0   
2       2.5+1.0     1.5+1.5     2.0+1.5      3.5+0.0      1.0+1.0     3.0+1.0   
3         NA+NA       NA+NA       NA+NA        NA+NA        NA+NA       NA+NA   
4       1.0+0.0     1.5+1.5     0.5+0.0     1.25+0.0       0.0+NB     2.5+0.0   
5       2.5+0.5     1.5+1.5     0.5+1.0     1.75+0.5      0.0+1.0     1.5+1.0   
6       3.5+1.0     1.5+0.5     1.5+0.5       US+1.0      0.5+1.0     2.0+1.0   
7       2.0+1.0     4.0+2.0     2.0+1.0      1.5+0.0      1.0+1.0     4.0+1.0   
8       0.5+0.0     1.0+0.0     0.5+1.0      1.0+0.0       0.5+NB       NB+NB   
9       1.0+0.5     0.0+1.5     0.5+0.5        NB+NB       0.0+NB       NB+NB   
10      2.5+1.0     3.0+2.0     1.0+0.5      1.0+0.5      1.0+0.5     2.0+1.0   

       Z9(4+1)    Z10(4+1)      Z11(4+1)              Z12(4+1)  \
0      3.5+0.5     2.0+0.0       0.5+1.0                 ZW+ZW   
1      3.5+1.0      0.0+NB       1.0+1.0     NS.(3.0)+NS.(1.0)   
2      3.5+1.0       ZW+ZW       1.0+0.5               3.5+1.0   
3        NA+NA       NA+NA         NA+NA                 NA+NA   
4        NB+NB       NB+NB         NB+NB                 NB+NB   
5      4.0+1.0       ZW+ZW       0.5+1.0               3.0+0.0   
6      3.5+1.0     2.0+0.0       0.5+0.5               3.0+1.0   
7      4.0+1.0     1.0+0.5        0.0+NB               3.5+1.0   
8        NB+NB       NB+NB         NB+NB                 NB+NB   
9        NB+NB       NB+NB         NB+NB                 NB+NB   
10     4.0+1.0     2.0+0.5       4.0+1.0               4.0+1.0   

                    Z13(5)  Z14(4+1) Z15(5)  
0                  4.0+0.0   4.0+1.0     NA  
1                  4.0+0.0   4.0+1.0     NA  
2                  4.0+1.0   4.0+1.0     NA  
3                    NA+NA     NA+NA     NA  
4                    NB+NB     NB+NB     NA  
5                  4.0+1.0   3.5+0.0     NA  
6                  2.0+1.0   4.0+1.0     NA  
7                  4.0+1.0   3.0+0.0     NA  
8                    NB+NB     NB+NB     NA  
9                    NB+NB     NB+NB     NA  
10                 2.0+0.0   4.0+1.0     NA

In [82]:
oceny[oceny_obj.columns] = oceny_obj.apply(lambda kolumna: kolumna.str.strip())

In [83]:
oceny.Nazwisko.values

array(['Nowak', 'Kowalski', 'Wiśniewski', 'Dąbrowski', 'Lewandowski',
       'Wójcik', 'Kamiński', 'Kowalczyk', 'Zielińska', 'Szymański',
       'Woźniak'], dtype=object)

In [80]:
oceny.drop(["int", "float"], axis=1, inplace=True)

## Eleganckie rozwiązanie (szybsze)

In [84]:
oceny = pd.read_csv("AiSD2.csv", skipinitialspace=True)
oceny.head()

Imię     Nazwisko  Z1(4+1)  Z2(4+1)  Z3(4+1)  Z4(4+1)  Z5(2+2)  \
0      Anna        Nowak  2.5+1.0  4.0+1.0  2.0+0.0  1.0+2.0  1.5+2.0   
1     Rafał     Kowalski  3.0+0.0  2.0+0.5  2.0+1.0  1.5+1.5  2.0+1.0   
2     Pavlo   Wiśniewski  2.5+0.5   US+1.0  2.5+1.0  1.5+1.5  2.0+1.5   
3    Aleksy    Dąbrowski  0.5+1.0  1.0+1.0    NA+NA    NA+NA    NA+NA   
4  Jarosław  Lewandowski  2.0+0.0  3.0+1.0  1.0+0.0  1.5+1.5  0.5+0.0   

    Z6(4+1)  Z7(4+1)  Z8(4+1)  Z9(4+1) Z10(4+1) Z11(4+1)           Z12(4+1)  \
0   2.0+0.0   0.0+NB  1.5+1.0  3.5+0.5  2.0+0.0  0.5+1.0              ZW+ZW   
1   1.0+0.5   0.0+NB  3.0+1.0  3.5+1.0   0.0+NB  1.0+1.0  NS.(3.0)+NS.(1.0)   
2   3.5+0.0  1.0+1.0  3.0+1.0  3.5+1.0    ZW+ZW  1.0+0.5            3.5+1.0   
3     NA+NA    NA+NA    NA+NA    NA+NA    NA+NA    NA+NA              NA+NA   
4  1.25+0.0   0.0+NB  2.5+0.0    NB+NB    NB+NB    NB+NB              NB+NB   

    Z13(5) Z14(4+1)  Z15(5)  
0  4.0+0.0  4.0+1.0     NaN  
1  4.0+0.0  4.0+1.0     NaN  
2  4.0+1.0  4.0+1.0     NaN  
3    NA+NA    NA+NA     NaN  
4    NB+NB    NB+NB     NaN

In [85]:
oceny.columns

Index(['Imię', 'Nazwisko', 'Z1(4+1)', 'Z2(4+1)', 'Z3(4+1)', 'Z4(4+1)',
       'Z5(2+2)', 'Z6(4+1)', 'Z7(4+1)', 'Z8(4+1)', 'Z9(4+1)', 'Z10(4+1)',
       'Z11(4+1)', 'Z12(4+1)', 'Z13(5)', 'Z14(4+1)', 'Z15(5)'],
      dtype='object')

In [86]:
oceny.Nazwisko.values

array(['Nowak', 'Kowalski', 'Wiśniewski', 'Dąbrowski', 'Lewandowski',
       'Wójcik', 'Kamiński', 'Kowalczyk', 'Zielińska', 'Szymański',
       'Woźniak'], dtype=object)

# E-maile

Załóżmy, że student nazywa się Łukasz Bączyński. Jego e-mail u nas na wydziale MiNI to
`baczynskil@student.mini.pw.edu.pl`

In [88]:
oceny.Imię[0]

'Anna'

In [89]:
oceny.Imię.str.slice(0, 1)

0     A
1     R
2     P
3     A
4     J
5     B
6     Ł
7     O
8     J
9     O
10    P
Name: Imię, dtype: object

In [90]:
oceny.Imię.str[0] # to polecam

0     A
1     R
2     P
3     A
4     J
5     B
6     Ł
7     O
8     J
9     O
10    P
Name: Imię, dtype: object

In [94]:
oceny.Imię.str[0:3]

0     Ann
1     Raf
2     Pav
3     Ale
4     Jar
5     Bar
6     Łuk
7     Osk
8     Joa
9     Okt
10    Pat
Name: Imię, dtype: object

In [95]:
oceny.Nazwisko.str.lower()

0           nowak
1        kowalski
2      wiśniewski
3       dąbrowski
4     lewandowski
5          wójcik
6        kamiński
7       kowalczyk
8       zielińska
9       szymański
10        woźniak
Name: Nazwisko, dtype: object

In [99]:
oceny['email'] = oceny.Nazwisko.str.lower()+ oceny.Imię.str[0:1].str.lower() + '@student.mini.pw.edu.pl'

In [100]:
oceny.head()

Imię     Nazwisko  Z1(4+1)  Z2(4+1)  Z3(4+1)  Z4(4+1)  Z5(2+2)  \
0      Anna        Nowak  2.5+1.0  4.0+1.0  2.0+0.0  1.0+2.0  1.5+2.0   
1     Rafał     Kowalski  3.0+0.0  2.0+0.5  2.0+1.0  1.5+1.5  2.0+1.0   
2     Pavlo   Wiśniewski  2.5+0.5   US+1.0  2.5+1.0  1.5+1.5  2.0+1.5   
3    Aleksy    Dąbrowski  0.5+1.0  1.0+1.0    NA+NA    NA+NA    NA+NA   
4  Jarosław  Lewandowski  2.0+0.0  3.0+1.0  1.0+0.0  1.5+1.5  0.5+0.0   

    Z6(4+1)  Z7(4+1)  Z8(4+1)  Z9(4+1) Z10(4+1) Z11(4+1)           Z12(4+1)  \
0   2.0+0.0   0.0+NB  1.5+1.0  3.5+0.5  2.0+0.0  0.5+1.0              ZW+ZW   
1   1.0+0.5   0.0+NB  3.0+1.0  3.5+1.0   0.0+NB  1.0+1.0  NS.(3.0)+NS.(1.0)   
2   3.5+0.0  1.0+1.0  3.0+1.0  3.5+1.0    ZW+ZW  1.0+0.5            3.5+1.0   
3     NA+NA    NA+NA    NA+NA    NA+NA    NA+NA    NA+NA              NA+NA   
4  1.25+0.0   0.0+NB  2.5+0.0    NB+NB    NB+NB    NB+NB              NB+NB   

    Z13(5) Z14(4+1)  Z15(5)                                email  
0  4.0+0.0  4.0+1.0     NaN        nowaka@student.mini.pw.edu.pl  
1  4.0+0.0  4.0+1.0     NaN     kowalskir@student.mini.pw.edu.pl  
2  4.0+1.0  4.0+1.0     NaN   wiśniewskip@student.mini.pw.edu.pl  
3    NA+NA    NA+NA     NaN    dąbrowskia@student.mini.pw.edu.pl  
4    NB+NB    NB+NB     NaN  lewandowskij@student.mini.pw.edu.pl

In [102]:
(oceny.Nazwisko + oceny.Imię.str[0] + "@student.mini.pw.edu.pl").str.lower()

0           nowaka@student.mini.pw.edu.pl
1        kowalskir@student.mini.pw.edu.pl
2      wiśniewskip@student.mini.pw.edu.pl
3       dąbrowskia@student.mini.pw.edu.pl
4     lewandowskij@student.mini.pw.edu.pl
5          wójcikb@student.mini.pw.edu.pl
6        kamińskił@student.mini.pw.edu.pl
7       kowalczyko@student.mini.pw.edu.pl
8       zielińskaj@student.mini.pw.edu.pl
9       szymańskio@student.mini.pw.edu.pl
10        woźniakp@student.mini.pw.edu.pl
dtype: object

!pip install unidecode

pip install unidecode

In [105]:
import unidecode

In [106]:
unidecode.unidecode("ąśćńółżźĄŚĆŻŹÓŃ")

'ascnolzzASCZZON'

In [107]:
unidecode.unidecode(oceny["email"])

AttributeError: 'Series' object has no attribute 'encode'

## Podejście 1, lepsze z apply z Pandasa

In [108]:
oceny["email"].apply(lambda x: print(x))

nowaka@student.mini.pw.edu.pl
kowalskir@student.mini.pw.edu.pl
wiśniewskip@student.mini.pw.edu.pl
dąbrowskia@student.mini.pw.edu.pl
lewandowskij@student.mini.pw.edu.pl
wójcikb@student.mini.pw.edu.pl
kamińskił@student.mini.pw.edu.pl
kowalczyko@student.mini.pw.edu.pl
zielińskaj@student.mini.pw.edu.pl
szymańskio@student.mini.pw.edu.pl
woźniakp@student.mini.pw.edu.pl


0     None
1     None
2     None
3     None
4     None
5     None
6     None
7     None
8     None
9     None
10    None
Name: email, dtype: object

In [109]:
oceny["email"].apply(lambda x: unidecode.unidecode(x))

0           nowaka@student.mini.pw.edu.pl
1        kowalskir@student.mini.pw.edu.pl
2      wisniewskip@student.mini.pw.edu.pl
3       dabrowskia@student.mini.pw.edu.pl
4     lewandowskij@student.mini.pw.edu.pl
5          wojcikb@student.mini.pw.edu.pl
6        kaminskil@student.mini.pw.edu.pl
7       kowalczyko@student.mini.pw.edu.pl
8       zielinskaj@student.mini.pw.edu.pl
9       szymanskio@student.mini.pw.edu.pl
10        wozniakp@student.mini.pw.edu.pl
Name: email, dtype: object

In [110]:
oceny["email"].apply(unidecode.unidecode)

0           nowaka@student.mini.pw.edu.pl
1        kowalskir@student.mini.pw.edu.pl
2      wisniewskip@student.mini.pw.edu.pl
3       dabrowskia@student.mini.pw.edu.pl
4     lewandowskij@student.mini.pw.edu.pl
5          wojcikb@student.mini.pw.edu.pl
6        kaminskil@student.mini.pw.edu.pl
7       kowalczyko@student.mini.pw.edu.pl
8       zielinskaj@student.mini.pw.edu.pl
9       szymanskio@student.mini.pw.edu.pl
10        wozniakp@student.mini.pw.edu.pl
Name: email, dtype: object

## Podejście 2, gorsze, z Numpy (ale też spoko)

In [111]:
import numpy as np

In [113]:
unidecode_zwektoryzowany = np.vectorize(unidecode.unidecode)

In [114]:
unidecode_zwektoryzowany(oceny["email"])

array(['nowaka@student.mini.pw.edu.pl',
       'kowalskir@student.mini.pw.edu.pl',
       'wisniewskip@student.mini.pw.edu.pl',
       'dabrowskia@student.mini.pw.edu.pl',
       'lewandowskij@student.mini.pw.edu.pl',
       'wojcikb@student.mini.pw.edu.pl',
       'kaminskil@student.mini.pw.edu.pl',
       'kowalczyko@student.mini.pw.edu.pl',
       'zielinskaj@student.mini.pw.edu.pl',
       'szymanskio@student.mini.pw.edu.pl',
       'wozniakp@student.mini.pw.edu.pl'], dtype='<U35')

In [115]:
oceny['email'] = unidecode_zwektoryzowany(oceny["email"])

In [116]:
oceny.head()

Imię     Nazwisko  Z1(4+1)  Z2(4+1)  Z3(4+1)  Z4(4+1)  Z5(2+2)  \
0      Anna        Nowak  2.5+1.0  4.0+1.0  2.0+0.0  1.0+2.0  1.5+2.0   
1     Rafał     Kowalski  3.0+0.0  2.0+0.5  2.0+1.0  1.5+1.5  2.0+1.0   
2     Pavlo   Wiśniewski  2.5+0.5   US+1.0  2.5+1.0  1.5+1.5  2.0+1.5   
3    Aleksy    Dąbrowski  0.5+1.0  1.0+1.0    NA+NA    NA+NA    NA+NA   
4  Jarosław  Lewandowski  2.0+0.0  3.0+1.0  1.0+0.0  1.5+1.5  0.5+0.0   

    Z6(4+1)  Z7(4+1)  Z8(4+1)  Z9(4+1) Z10(4+1) Z11(4+1)           Z12(4+1)  \
0   2.0+0.0   0.0+NB  1.5+1.0  3.5+0.5  2.0+0.0  0.5+1.0              ZW+ZW   
1   1.0+0.5   0.0+NB  3.0+1.0  3.5+1.0   0.0+NB  1.0+1.0  NS.(3.0)+NS.(1.0)   
2   3.5+0.0  1.0+1.0  3.0+1.0  3.5+1.0    ZW+ZW  1.0+0.5            3.5+1.0   
3     NA+NA    NA+NA    NA+NA    NA+NA    NA+NA    NA+NA              NA+NA   
4  1.25+0.0   0.0+NB  2.5+0.0    NB+NB    NB+NB    NB+NB              NB+NB   

    Z13(5) Z14(4+1)  Z15(5)                                email  
0  4.0+0.0  4.0+1.0     NaN        nowaka@student.mini.pw.edu.pl  
1  4.0+0.0  4.0+1.0     NaN     kowalskir@student.mini.pw.edu.pl  
2  4.0+1.0  4.0+1.0     NaN   wisniewskip@student.mini.pw.edu.pl  
3    NA+NA    NA+NA     NaN    dabrowskia@student.mini.pw.edu.pl  
4    NB+NB    NB+NB     NaN  lewandowskij@student.mini.pw.edu.pl

## E-mail jako jeden napis

Chcę połączyć wszystkie maila w taki napis: "student1@domena.pl, student2@domena.pl, student3@domena.pl, ..."

* odpowiednia metoda str w Pandasie
* operacje na napisach

### operacje na napisach

In [122]:
l = ["element1", "element2", "element2"]
'#separator#'.join(l)

'element1#separator#element2#separator#element2'

In [121]:
', '.join(oceny['email'].values)

'nowaka@student.mini.pw.edu.pl, kowalskir@student.mini.pw.edu.pl, wisniewskip@student.mini.pw.edu.pl, dabrowskia@student.mini.pw.edu.pl, lewandowskij@student.mini.pw.edu.pl, wojcikb@student.mini.pw.edu.pl, kaminskil@student.mini.pw.edu.pl, kowalczyko@student.mini.pw.edu.pl, zielinskaj@student.mini.pw.edu.pl, szymanskio@student.mini.pw.edu.pl, wozniakp@student.mini.pw.edu.pl'

### metoda w Pandasie

In [123]:
oceny["email"].str.cat(sep=", ")

'nowaka@student.mini.pw.edu.pl, kowalskir@student.mini.pw.edu.pl, wisniewskip@student.mini.pw.edu.pl, dabrowskia@student.mini.pw.edu.pl, lewandowskij@student.mini.pw.edu.pl, wojcikb@student.mini.pw.edu.pl, kaminskil@student.mini.pw.edu.pl, kowalczyko@student.mini.pw.edu.pl, zielinskaj@student.mini.pw.edu.pl, szymanskio@student.mini.pw.edu.pl, wozniakp@student.mini.pw.edu.pl'

# Rozbij kolumny na część labolatoryjną i domową

In [125]:
oceny.head()

Imię     Nazwisko  Z1(4+1)  Z2(4+1)  Z3(4+1)  Z4(4+1)  Z5(2+2)  \
0      Anna        Nowak  2.5+1.0  4.0+1.0  2.0+0.0  1.0+2.0  1.5+2.0   
1     Rafał     Kowalski  3.0+0.0  2.0+0.5  2.0+1.0  1.5+1.5  2.0+1.0   
2     Pavlo   Wiśniewski  2.5+0.5   US+1.0  2.5+1.0  1.5+1.5  2.0+1.5   
3    Aleksy    Dąbrowski  0.5+1.0  1.0+1.0    NA+NA    NA+NA    NA+NA   
4  Jarosław  Lewandowski  2.0+0.0  3.0+1.0  1.0+0.0  1.5+1.5  0.5+0.0   

    Z6(4+1)  Z7(4+1)  Z8(4+1)  Z9(4+1) Z10(4+1) Z11(4+1)           Z12(4+1)  \
0   2.0+0.0   0.0+NB  1.5+1.0  3.5+0.5  2.0+0.0  0.5+1.0              ZW+ZW   
1   1.0+0.5   0.0+NB  3.0+1.0  3.5+1.0   0.0+NB  1.0+1.0  NS.(3.0)+NS.(1.0)   
2   3.5+0.0  1.0+1.0  3.0+1.0  3.5+1.0    ZW+ZW  1.0+0.5            3.5+1.0   
3     NA+NA    NA+NA    NA+NA    NA+NA    NA+NA    NA+NA              NA+NA   
4  1.25+0.0   0.0+NB  2.5+0.0    NB+NB    NB+NB    NB+NB              NB+NB   

    Z13(5) Z14(4+1)  Z15(5)                                email  
0  4.0+0.0  4.0+1.0     NaN        nowaka@student.mini.pw.edu.pl  
1  4.0+0.0  4.0+1.0     NaN     kowalskir@student.mini.pw.edu.pl  
2  4.0+1.0  4.0+1.0     NaN   wisniewskip@student.mini.pw.edu.pl  
3    NA+NA    NA+NA     NaN    dabrowskia@student.mini.pw.edu.pl  
4    NB+NB    NB+NB     NaN  lewandowskij@student.mini.pw.edu.pl

Jeśli mam kolumnę Z1(4+1), to chciałbym mieć dwie kolumny: `L1(4)` oraz '`D1(1)` i oczywiście wartości w kolumnie odpowiednio rozbite.

Trzeba pamiętać żeby nie przetowrzyć kolumn: Imię, Nazwisko, email

* `\d`- cyfra
* `()` - grupa (capturing group)
* `\( \)` - dosłowne rozumienie znaku nawias

In [126]:
import re

In [128]:
re.findall("[AO]la", "Ala ma kota. Ola ma psa.")

['Ala', 'Ola']

Z1(4+1)

In [132]:
re.findall("Z(\d{1,2})\((\d)\+(\d)\)", "Z1(4+1)")

[('1', '4', '1')]

In [133]:
re.findall("Z(\d{1,2})\((\d)\+(\d)\)", "Kowalski")

[]

In [134]:
"2+3".split("+")

['2', '3']

In [138]:
oceny["Z1(4+1)"].str.split("+", expand=True)

0    1
0   2.5  1.0
1   3.0  0.0
2   2.5  0.5
3   0.5  1.0
4   2.0  0.0
5   2.0  1.0
6   4.0  1.0
7   4.0  0.5
8   2.5  0.0
9   2.5  1.0
10  2.0  1.0

In [139]:
oceny2 = oceny.copy()

In [140]:
# najpierw zmienamiamy nazwe kolumny Z13(5) na Z13(4+1)
# usuwamy Z15(5)

# for po nazwach kolumn:
    # sprawdzamy dopasowanie wzorca nazwy kolumny
    # jesli dopasowano (len(result) > 0)
        # bierzemy te grupy
        # bierzemy rozbite wartości
        # tworzymy dwie kolumny (L oraz D)
        # drop kolumny obecnie przetwarzanej

In [141]:
oceny2.rename(columns={"Z13(5)": "Z13(4+1)"}, inplace=True)

In [143]:
oceny2.drop(columns="Z15(5)", axis=1, inplace=True)

In [146]:
oceny2.columns

Index(['Imię', 'Nazwisko', 'Z1(4+1)', 'Z2(4+1)', 'Z3(4+1)', 'Z4(4+1)',
       'Z5(2+2)', 'Z6(4+1)', 'Z7(4+1)', 'Z8(4+1)', 'Z9(4+1)', 'Z10(4+1)',
       'Z11(4+1)', 'Z12(4+1)', 'Z13(4+1)', 'Z14(4+1)', 'email'],
      dtype='object')

In [149]:
for i in [1,2,3,4,5]:
    if i == 3:
        continue
    print(i)

1
2
4
5


In [151]:
"L" + "2" + "(" + "3" + ")"

'L2(3)'

In [152]:
f"L{2}({3})"

'L2(3)'

In [153]:
for kolumna in oceny2.columns:
    dopasowanie_wzorca = re.findall("Z(\d{1,2})\((\d)\+(\d)\)", kolumna)
    if len(dopasowanie_wzorca) == 0:
        continue
    numer, maks_lab, maks_dom = dopasowanie_wzorca[0]
    rozbite_wartosci = oceny2[kolumna].str.split("+", expand=True)
    oceny2[f"L{numer}({maks_lab})"] = rozbite_wartosci[0]
    oceny2[f"D{numer}({maks_dom})"] = rozbite_wartosci[1]
    oceny2.drop(columns=kolumna, inplace=True)

In [154]:
oceny2.head()

Imię     Nazwisko                                email L1(4) D1(1)  \
0      Anna        Nowak        nowaka@student.mini.pw.edu.pl   2.5   1.0   
1     Rafał     Kowalski     kowalskir@student.mini.pw.edu.pl   3.0   0.0   
2     Pavlo   Wiśniewski   wisniewskip@student.mini.pw.edu.pl   2.5   0.5   
3    Aleksy    Dąbrowski    dabrowskia@student.mini.pw.edu.pl   0.5   1.0   
4  Jarosław  Lewandowski  lewandowskij@student.mini.pw.edu.pl   2.0   0.0   

  L2(4) D2(1) L3(4) D3(1) L4(4)  ... L10(4) D10(1) L11(4) D11(1)    L12(4)  \
0   4.0   1.0   2.0   0.0   1.0  ...    2.0    0.0    0.5    1.0        ZW   
1   2.0   0.5   2.0   1.0   1.5  ...    0.0     NB    1.0    1.0  NS.(3.0)   
2    US   1.0   2.5   1.0   1.5  ...     ZW     ZW    1.0    0.5       3.5   
3   1.0   1.0    NA    NA    NA  ...     NA     NA     NA     NA        NA   
4   3.0   1.0   1.0   0.0   1.5  ...     NB     NB     NB     NB        NB   

     D12(1) L13(4) D13(1) L14(4) D14(1)  
0        ZW    4.0    0.0    4.0    1.0  
1  NS.(1.0)    4.0    0.0    4.0    1.0  
2       1.0    4.0    1.0    4.0    1.0  
3        NA     NA     NA     NA     NA  
4        NB     NB     NB     NB     NB  

[5 rows x 31 columns]

# Zamieniamy nieobecności/plagiaty

* NA, NB - nieobecność nieusprawiedliwiona, wstawiamy 0
* US, ZW - nieobecność usprawiedliwiona, wstawiamy np.NaN (o co chodzi - o tym za chwilę)
* NS.(1.0) - praca niesamodzielna, wstawiamy 0

# Wstawiamy punkty za zwolnienia lekarskie

* Zamieniamy kolumny na floaty
* Dla każdego znalezionego NaN: policz średnią z wiersza (siła studenta), średnią z kolumny (trudność zadania), oblicz średnią z tych dwóch średnich i wstaw (zrobić pętlą, ja potem pokażę w Numpy)

### Wstawienie punktów za usprawiedliwione nieobecności (wersja z pętlami)

In [71]:
# pętla po wierszach:
#    pętla po kolumnach (odpowiednich):
#        czy napotkałem nan?
#            liczę średnie, średnią z średnich, wstawiam

### Wstawienie punktów za usprawiedliwione nieobecności (wersja zwektoryzowana)

In [75]:
# policzyć średnie dla każdej kolumny i każdego wiersza (w sposób zwektoryzowany)
# znaleźć indeksy, na których są nany (zwektoryzowany)
# podstawić pod te indeksy średnie z średnich (indeksowanie listą liczb całkowitych)

# Wystawiamy oceny (nareszcie!)

* Obliczyć teoretyczną maksymalną sumę punktów, jaką można było uzyskać (przejść po nazwach kolumn, odzyskać maks. punkty, zsumować)
* Zsumować punkty studentów (nowa kolumna - Suma)
* Obliczyć procent uzyskanych punktów. Przelicznik jest następujący:

* [0, 50%) -> 2.0
* [50%, 60%) -> 3.0
* [60%, 70%) -> 3.5
* [70%, 80%) -> 4.0
* [80%, 90%) -> 4.5
* [90%, 100%] -> 5.0


Pomocna metoda: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.cut.html